relion/convert/convert31.py

```diff
@@ -589,22 +589,73 @@ def _rowToPart(self, row, particle):
 
         # TODO: coord extra labels, partId, micId,
         if self._setCoord:
-            coord = self.rowToCoord(row)
+            coord = Coordinate()
+            self.rowToCoord(row, coord)
             particle.setCoordinate(coord)
 
         if self._postprocessImageRow:
             self._postprocessImageRow(particle, row)
 
+    def readSetOfCoordinates(self, starFile, coordSet, micList=None, **kwargs):
+        """ Convert a star file into a set of coordinates.
+
+        Params:
+            starFile: the filename of the star file
+            coordSet: output coordinates set
+            micList: list of micNames to match coordSet
+
+        Keyword Arguments:
+            postprocessCoordRow:
+            extraLabels:
+
+        """
+        self._postprocessCoordRow = kwargs.get('postprocessCoordRow', None)
+        coordsReader = Table.Reader(starFile, tableName='', types=LABELS_DICT)
+        if not coordsReader.hasAllColumns(self.COORD_LABELS[:3]):
+            raise Exception("STAR file should include columns: ", self.COORD_LABELS[:3])
+
+        coordsReader = sorted(coordsReader, key=lambda r: getattr(r, 'rlnMicrographName'))
+        coordsReader = [row for row in coordsReader if row.rlnMicrographName in micList]
+
+        firstRow = coordsReader[0]
+        hasMicId = firstRow.get('rlnMicrographId', False)
+
+        coord = Coordinate()
+        self.rowToCoord(firstRow, coord)
+        if hasMicId:
```

Comment:
Why the first coordinate is treated separated from the rest in the loop (also duplicating some code)

Reply:
Because of self.createExtraLabels(coord, firstRow, extraLabels) that needs to create extra objects first